Source/Particles/Deposition/ChargeDeposition.H

```diff
@@ -201,6 +201,323 @@ void doChargeDepositionShapeN (const GetParticlePosition& GetPosition,
+    amrex::ignore_unused(np_to_deposit);
+    // Loop with one block per tile (the shared memory is allocated on a per-block basis)
+    // The threads within each block loop over the particles of its tile
+    // (Each threads processes a different set of particles.)
+    amrex::launch(
+                  nblocks, threads_per_block, shared_mem_bytes, amrex::Gpu::gpuStream(),
+                  [=] AMREX_GPU_DEVICE () noexcept
+#else // defined(AMREX_USE_CUDA) || defined(AMREX_USE_HIP)
+    amrex::ParallelFor(np_to_deposit, [=] AMREX_GPU_DEVICE (long ip_orig) noexcept
+#endif
+      {
+#if defined(AMREX_USE_CUDA) || defined(AMREX_USE_HIP)
+        const int bin_id = blockIdx.x;
+        const unsigned int bin_start = offsets_ptr[bin_id];
+        const unsigned int bin_stop = offsets_ptr[bin_id+1];
+
+        if (bin_start == bin_stop) { return; }
+
+        amrex::Box buffer_box;
+        {
+          ParticleReal xp, yp, zp;
+          GetPosition(permutation[bin_start], xp, yp, zp);
+#if defined(WARPX_DIM_3D)
+          IntVect iv = IntVect(int(amrex::Math::floor((xp-plo[0])*dxiarr[0])),
+                               int(amrex::Math::floor((yp-plo[1])*dxiarr[1])),
+                               int(amrex::Math::floor((zp-plo[2])*dxiarr[2])));
+#elif defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+          IntVect iv = IntVect(
+                               int(amrex::Math::floor((xp-plo[0])*dxiarr[0])),
+                               int(amrex::Math::floor((zp-plo[1])*dxiarr[1])));
+#elif defined(WARPX_DIM_1D_Z)
+          IntVect iv = IntVect(int(amrex::Math::floor((zp-plo[0])*dxiarr[0])));
+#endif
+          iv += domain.smallEnd();
+          getTileIndex(iv, box, true, bin_size, buffer_box);
+        }
+
+        Box tbx = convert( buffer_box, ix_type);
+        tbx.grow(depos_order);
+
+        Gpu::SharedMemory<amrex::Real> gsm;
+        amrex::Real* const shared = gsm.dataPtr();
+
+        amrex::Array4<amrex::Real> buf(shared, amrex::begin(tbx), amrex::end(tbx), 1);
+
+        // Zero-initialize the temporary array in shared memory
+        volatile amrex::Real* vs = shared;
+        for (int i = threadIdx.x; i < tbx.numPts(); i += blockDim.x) {
+          vs[i] = 0.0;
+        }
+        __syncthreads();
+#else
+        amrex::Array4<amrex::Real> const &buf = rho_arr;
+#endif // defined(AMREX_USE_CUDA) || defined(AMREX_USE_HIP)
+
+#if defined(AMREX_USE_CUDA) || defined(AMREX_USE_HIP)
+        // Loop over macroparticles: each threads loops over particles with a stride of `blockDim.x`
+        for (unsigned int ip_orig = bin_start + threadIdx.x; ip_orig < bin_stop; ip_orig += blockDim.x)
+#endif
+        {
+            unsigned int ip = permutation[ip_orig];
+
+#if defined(WARPX_USE_GPUCLOCK)
+            KernelTimer kernelTimer(cost && load_balance_costs_update_algo
+                                 == LoadBalanceCostsUpdateAlgo::GpuClock, cost_real);
+#endif
+            // --- Get particle quantities
```

Comment:
It might be a good idea to refactor from here down to extract a separate function `depositChargeSingleParticleShapeN()`, since that's common across the two different implementations. The only difference is the actual array targeted by the atomic additions

Reply:
As I go to extract it, I realize that this may have implications for GPU register allocation that it would need to be careful of.